### DISCUSSION TITLE
How is mutation of passed variable possible?

### DISCUSSION THREAD

**Discussion by Chappie74** [flame-engine/flame#1983] [Q&A] [answered]
Hi I'm new to Flame and I am currently going through the Klondike tutorial. 

I've noticed that modifying the properties of a component that exists in the World space, passed via a function, mutates the actual object of the World space. (sorry if its confusing)

Take for example 
![image](https://user-images.githubusercontent.com/21108487/193437728-5ca71f3f-b308-4d54-acfc-56a9ce25418a.png)

On line 70 we iterate through all the cards and call the `acquireCard` function of the stock component. Seems good so far. 

My confusion comes from what the `acquireCard`  does. From my understanding dart is pass by value. So how does mutating the properties(the position mainly) of the passed card object able to affect the card defined in the world space?
![image](https://user-images.githubusercontent.com/21108487/193437781-cd1b1971-5883-48b3-b611-9e9c5b1aa0f7.png)

I'm more expecting to have to update/replace the list of cards in the world space with the newly updated card, rather than just be able to mutate it in the Stock component and just have it work. Am I missing something?
If someone could please explain why this is possible that would be real helpful. THanks in advance.

**Comment by st-pasha** [comment, marked as answer]
In Dart, only primitive types are passed by values: booleans, integers, doubles. Every other object is passed "by reference". For example, if I have a `card` object, and pass it to `someFunction(card)`, then the function will see exactly the same object as known to the caller. If this was C++, then the semantics would be expressed as `auto someFunction(Card& card)`.

If `someFunction` modifies the `card` object, those changes will be visible to the caller -- because it's just the same object.

In Dart there is no notion of passing objects "by value" -- if you want to create an object copy, you would have to manually define and call the `Card.clone()` method. Neither there is a way to pass an object by "constant reference".

**Reply by Chappie74** [reply]
Damn I did not know this. Thank you very much.
